ql/src/java/org/apache/hadoop/hive/ql/QueryCompilationSummaryHook.java

```diff
@@ -0,0 +1,131 @@
+
+    }
+
+    private static final class CompileStep {
+      private String name;
+      private List<CompileStep> subSteps;
+
+      public CompileStep(String name) {
+        this.name = name;
+        subSteps = new ArrayList<>();
+      }
+      public void addSubStep(CompileStep subStep) {
+        subSteps.add(subStep);
+      }
+      public List<CompileStep> getSubSteps() {
+          return subSteps;
+      }
+    }
+
+    static {
+      compileSteps.add(new CompileStep(PerfLogger.PARSE));
+      compileSteps.add(new CompileStep(PerfLogger.GENERATE_RESOLVED_PARSETREE));
+      CompileStep logicalPlanAndOpTree = new CompileStep(PerfLogger.LOGICALPLAN_AND_HIVE_OPERATOR_TREE);
+      CompileStep logicalPlan = new CompileStep(PerfLogger.GENERATE_LOGICAL_PLAN);
+      logicalPlanAndOpTree.addSubStep(logicalPlan);
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.PLAN_GENERATION));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.MV_REWRITE_FIELD_TRIMMER));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.REMOVING_SUBQUERY));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.DECORRELATION));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.VALIDATE_QUERY_MATERIALIZATION));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.PREJOIN_ORDERING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.MV_REWRITING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.JOIN_REORDERING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.POSTJOIN_ORDERING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.HIVE_SORT_PREDICATES));
+      logicalPlanAndOpTree.addSubStep(new CompileStep(PerfLogger.GENERATE_OPERATOR_TREE));
+      compileSteps.add(logicalPlanAndOpTree);
+      compileSteps.add(new CompileStep(PerfLogger.DEDUCE_RESULTSET_SCHEMA));
+      compileSteps.add(new CompileStep(PerfLogger.PARSE_CONTEXT_GENERATION));
+      compileSteps.add(new CompileStep(PerfLogger.SAVE_AND_VALIDATE_VIEW));
+      compileSteps.add(new CompileStep(PerfLogger.LOGICAL_OPTIMIZATION));
+      compileSteps.add(new CompileStep(PerfLogger.PHYSICAL_OPTIMIZATION));
+      compileSteps.add(new CompileStep(PerfLogger.POST_PROCESSING));
+    }
+
+    public void getCompileSummary() {
+      StringBuilder compileSummary = new StringBuilder();
+      compileSummary.append(QUERY_COMPILATION_SUMMARY);
+      compileSummary.append(System.getProperty(LINE_SEPARATOR));
+      compileSummary.append(SEPARATOR);
+      compileSummary.append(System.getProperty(LINE_SEPARATOR));
+      PerfLogger perfLogger = SessionState.getPerfLogger();
+      appendCompileSteps(compileSummary, perfLogger, "" ,1, compileSteps);
+      compileSummary.append(format(TOTAL_COMPILATION_TIME, perfLogger.getDuration(PerfLogger.COMPILE)));
+      compileSummary.append(System.getProperty(LINE_SEPARATOR));
+      compileSummary.append(SEPARATOR);
+      SessionState.getConsole().printInfo(compileSummary.toString());
```

Comment:
Maybe we should use `printInfo(summary, isSilent=false)` to make the summary appear in `q.out` or something along these lines.